### DISCUSSION TITLE
Target Sensor Trigger

### DISCUSSION THREAD

**Discussion by Markyt77** [BottlecapDave/HomeAssistant-OctopusEnergy#399] [General]
I'm getting close to finishing a project that takes my forecast solar, desired battery SOC, estimated usage etc and tries to work out whether to charge the batteries from the mains to make up any predicted solar shortfall. I have an automation that works out the potential energy deficit then updates the number of slots required in my target sensor. This automation is triggered when the solar forecast for today changes (which is often thanks to British weather)

I started with the target sensors set to evaluate multiple times a day but it gave me unexpected results (see issue #359) Having the target sensor update once a day means that when forecast solar drops off a cliff the only way I can see to update the target slots is to edit and re-save the target sensor.

Is there a way to trigger that target sensor evaluation so I can force it to update as per my my time slot automation? Or am I approaching this wrong?

**Comment by BottlecapDave** [comment]
Hi there. You should be able to take advantage of the [update_target_config](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_target_rate.md#service-octopus_energyupdate_target_config) service to dynamically update the start/end times the target rate sensor should find the best rates for. Specifically, you probably want to set the start time to "now". You can also use this to update the target number of hours based on your calculations. 

When this service is called, this will cause an evaluation betwen the two new times and for the potentially new target hours.

Hope this helps.

**Comment by Markyt77** [comment]
Thanks David, that makes sense, so you mean something like this? 

    - service: octopus_energy.update_target_config
      data:
        target_hours: >
          "{{ states('input_number.octopus_energy_target_hours') | string }}"
        target_start_time: >
          {{'%02d'% now().hour}}:{{'%02d'% now().minute}}
        target_end_time: >
         {{'%02d'% now().hour}}:{{'%02d'% now().minute}}

**Reply by BottlecapDave** [reply]
It depends on how specific you want to be. If you are already working out how many hours you need, then you don't need to include `target_start_time` and `target_end_time` and just update the hours accordingly. However if you do this, there's a chance the chosen timeframe could be in the past, and therefore will evaluate to a time for the following day. You _could_ reduce this chance by finding [the last applicable times](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_target_rate.md#latest-period) within your timeframe, but again this mileage may vary on agile tariffs.

A way to get around is to set just the start_time to `now` which will force it to find the best time in the present. You could do this, or also set the target hours dynamically.

```
- service: octopus_energy.update_target_config
  data:
    target_start_time: >
      {{'%02d'% now().hour}}:{{'%02d'% now().minute}}
```

This will mean the target timeframe will constantly shift. The downside to this approach is you might get failures if the start/end time is smaller than your desired hours which you might want to account for.

If you do what you suggested, there's a chance that the target sensor might not turn on until the next day (because the start/end time spans multiple days).

**Reply by Markyt77** [reply]
Hi Dave,thanks for getting back to me. I am indeed seeing the behaviour you describe. Today's forecast has gone from 9kw to 2kw and whilst the target update is mostly working, it's also picking out slots in the past. I've run a few tests by manually updating the target rate sensor but it only accepts 16:00 or 23:00 as per the docs. I'm not going to make it cranky by forcing now() on the back end when the front end only accepts agile hours am I?

**Reply by Markyt77** [reply]
Update, using now() is simply rejected with the following in traces > Target time not fit for agile tariffs. Please consult target rate documentation for more information.

**Reply by BottlecapDave** [reply]
Sorry for the very late response. Did you solve this issue?

The target rate sensor for agile can be set to times other than 16:00 or 23:00, however if you set the start time to before 16:00 then you can't set the end time to after 23:00 as the target rate sensor will have incomplete data which will cause weird results in some scenarios.

Therefore you'll need to do a bit of juggling which again may or may not get your desired results.

```yaml
- service: octopus_energy.update_target_config
  data:
    target_start_time: >
      {{'%02d'% now().hour}}:{{'%02d'% now().minute}}
    target_end_time: >
     {% if now().hour < 16 %}
       23:00
     {% else %}
       {{'%02d'% now().hour}}:{{'%02d'% now().minute}}
     {% endif %}
```

This will try and ensure that the target period starts now and looks as far into the future as possible. If we're before 4pm then the furthest into the future we can look is 11pm as this is what agile rates go up to. If we're after 4pm then we can look a full 24 hours ahead.

The issue with this though is that your target times could fail if you're requesting more hours than the desired time frame (e.g. you're asking for 9 hours and your start/end time is 15:00-23:00). For this you'll need to have some additional logic around `target_hours`

**Reply by Markyt77** [reply]
Kinda, I solved it by buying an extra storage battery to make up for solcast getting confused by 'clouds' :) I've just tried your code and it works perfectly, thanks so much! As you say, it'll need some extra logic regarding hours but I now have what I need to move forward, thanks again

**Comment by Markyt77** [comment]
I just tried it, setting target_start_time to now() does actually work, but as target_end_time was still set to 23:00 it only picked up tonight's more expensive rates, not the cheap slots at 4am. Setting target_start_time & target_end_time to now() seems to make it throw up and return nothing at all.
